### PR TITLE
Development notes

### PR DESCRIPTION
**Pull request by mohit**
Add notes on how to test workflows and an overview of the important packages.

### CODE REVIEW

File: handler.go [hunk 1]
```diff
@@ -101,11 +101,6 @@ func (wm WorkflowManager) StartJobForWorkflow(ctx context.Context, input *models
 		return &models.Job{}, err
 	}
 
-	err = wm.store.SaveJob(*job)
```
Comment (mohit) [inline handler.go]: This is a bugfix that really should be part of another PR but 🤷‍♂️

File: README.md [hunk 2]
```diff
@@ -4,15 +4,94 @@ Minimal Workflow orchestrator for AWS Batch
...
+
+Also contains `KitchenSink` which is a workflow resource that can be used by other packages in tests. 
+
+* `executor`: contains JobManager and a wrapper over `aws-go-sdk`'s Batch API
+  * Yes this needs better naming but this is just easier to work with now, and should be easy to refactor.
+  * Suggestions for better naming and organization welcome!
+  * `JobManager` currently is currently a place where Workflows are converted to tasks and vice versa.
+  *  `JobManager` also converts between the resources to details that can be consumed by the `BatchClient` 
+
+* `store` defines and interface for the store and contains a very simple in-memory implementation
```
Comment (nathanleiby) [inline README.md]: s/and/an

File: README.md [hunk 3]
```diff
@@ -4,15 +4,94 @@ Minimal Workflow orchestrator for AWS Batch
...
+
+### Setting up a workflow
+
+1. Setup a Batch environment in AWS with a `JobQueue` and a `ComputeEnvironment`. 
+
+2. Create one or more `JobDefinitions` the `AWS CLI` or [UI](https://console.aws.amazon.com/batch/home?region=us-east-1#/job-definitions/new) (with IAM Permissions)
```
Comment (nathanleiby) [inline README.md]: s/the/via the

Comment (nathanleiby) [inline README.md]: Add link to aws cli?

File: README.md [hunk 4]
```diff
@@ -4,15 +4,94 @@ Minimal Workflow orchestrator for AWS Batch
...
+
+3. Your app role should have the `batchcli-dev` policy (defined below) appended to it
```
Comment (nathanleiby) [inline README.md]: maybe put the policy in a file, and then link to that file

File: README.md [hunk 5]
```diff
@@ -4,15 +4,94 @@ Minimal Workflow orchestrator for AWS Batch
...
+* `main`: contains the api `handler` 
 
-- Run `make build`, `make run`, or `make test` - This should fail with an error about having to implement the business logic.
+* `resources`: structs for the major concepts used by the `workflow-manager` namely
 
-- Implement aforementioned business logic so that code will build
+  * WorkflowDefinition
+  * State
+  * Job
+  * Task
```
Comment (nathanleiby) [inline README.md]: generally: Given that this is an open-source project, I think we should name things to align with AWS names as much as possible.

If we prefer to have our control plane at Clever call them something different, that's fine -- most people won't be interacting with this repo

File: README.md [hunk 6]
```diff
@@ -4,15 +4,94 @@ Minimal Workflow orchestrator for AWS Batch
...
+
+2. Create one or more `JobDefinitions` the `AWS CLI` or [UI](https://console.aws.amazon.com/batch/home?region=us-east-1#/job-definitions/new) (with IAM Permissions)
+
+3. Your app role should have the `batchcli-dev` policy (defined below) appended to it
+
+4. Define a workflow json (see [example](./data/hello-workflow.post.json) with the resource as the `job-definition-name:version-number`
+
+### Running a job
+
+0. Run workflow-manager on your local machine or on an ECS cluster.
```
Comment (nathanleiby) [inline README.md]: nit: how? `make run`.. `ark start --local`..

File: README.md [hunk 7]
```diff
@@ -4,15 +4,94 @@ Minimal Workflow orchestrator for AWS Batch
 
 Owned by eng-infra
 
-## Developing
+## Development
 
-- Update swagger.yml with your endpoints. See the [Swagger spec](http://swagger.io/specification/) for additional details on defining your swagger file.
+### Overview of packages
```
Comment (nathanleiby) [inline README.md]: nit: in long run, could just use auto-generated Go docs to describe packages

File: README.md [hunk 8]
```diff
@@ -4,15 +4,94 @@ Minimal Workflow orchestrator for AWS Batch
...
+    "data": [
+      "arg1",
+      "arg2",
+      "arg3"
+    ]
+  }'
+```
+
+4. Monitor the status of the job using
+
```
Comment (nathanleiby) [inline README.md]: could document the difference between these two commands, which I think is

1. to monitor a single job
2. to monitor the status of all jobs in the current workflow

File: README.md [hunk 9]
```diff
@@ -4,15 +4,94 @@ Minimal Workflow orchestrator for AWS Batch
...
+0. Run workflow-manager on your local machine or on an ECS cluster.
+
+1. Post the workflow using the API
+
+```sh
+curl -XPOST http://localhost:8080/workflows -d @data/hello-workflow.post.json
+```
+
+2. [optional]  Test the above worked.
```
Comment (nathanleiby) [inline README.md]: s/Test the above worked/Verify the workflow was created

File: README.md (hunk 5 above)
Comment (nathanleiby) [inline README.md]: To me this means:

- Workflow -> WorkflowDefinition
- Job -> Workflow

This avoids conflating our notion of `Job` with Batch's `Job`

Comment (nathanleiby) [inline README.md]: This naming also parallels how things in AWS often have a definition and an instance.

Then, we can think of Workflow manager as just adding two new primitives beyond those provided by AWS.

Comment (nathanleiby) [inline README.md]: (I don't want to 🚲 🏠 , and I know you've thought a lot about this)

Comment (mohit) [inline README.md]: I did ask everyone about thoughts on this but did not get any opinions. This was a slightly arbitrary choice I made. I'll probably make the change tomorrow after another discussion. Keeping it the same for now.

Comment (nathanleiby) [inline README.md]: 👍 sounds good!
